### ISSUE TITLE
[Bug] Connection Leaks in manage.py, IndexOperationsApi, VectorOperationsApi

### ISSUE THREAD

**Issue by rbuck-som** [pinecone-io/pinecone-python-client#230]
### Is this a new bug in the Pinecone Python client?

- [X] I believe this is a new bug in the Pinecone Python Client
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Your python client has memory leaks that proved problematic...

Top issue is this: you leak **5 descriptors per API call**, never to be reclaimed until eventually the Python process inevitably and eventually crashes because it exceeds max open file descriptors. Why does this happen? My answers are below...

Sample output from `lsof` to point out the problem; this is one of 32,000 file descriptor leaks created in relative short order using your product:

```
.... 10-20 thousand more above...
Python 60130 robert.buck 7221u IPv4 0xd946e14d9f16b61b 0t0 TCP 192.168.0.103:60745->ec2-18-213-200-10.compute-1.amazonaws.com:https (ESTABLISHED)
Python 60130 robert.buck 7222 PIPE 0x2727afc4fb51e58a 16384 ->0xd97beed562c859f2
Python 60130 robert.buck 7223 PIPE 0xd97beed562c859f2 16384 ->0x2727afc4fb51e58a
Python 60130 robert.buck 7224r PSXSEM 0t0 /mp-r4wqhiyk
Python 60130 robert.buck 7225r PSXSEM 0t0 /mp-vpwtcej6
.... 10-20 thousand more below...
```

I didn't know if it was your product or another product. So...

I instrumented the socket.socket API in Python only to find out what's going on. I created a command line tool whose sole purpose is to crush your API to see how it breaks. It didn't take but seconds to break your API. This is (I suspect) related to other support events that were side-stepped by Pinecone.io support, related to service issues in AWS. I could trace the issues back to your product via some socket instrumentation:

```
def generate_stack_trace():
    stack = traceback.extract_stack()
    print(f"\nStack Trace:\n")
    for frame in stack:
        filename, line_number, function_name, code = frame
        print(f"File: {filename}, Line: {line_number}, Function: {function_name}, Code: {code}")
    print(f"\n-------\n\n")

class TraceSocket(socket.socket):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        logging.info(f"socket created: {self.getsockname()}")
        generate_stack_trace()

    def close(self) -> None:
        logging.info(f"socket closed: {self.getsockname()}")
        super().close()


socket.socket = TraceSocket
```

Analysis...

In a nutshell, your code in `rest.py`, `api_client.py,` all go through great and appropriate lengths to ensure that connections can be reused and cleaned up when a client is disposed. But, that's where the goodness ends. Other parts have leaks... The rest of the code walks away from even taking advantage of Python's `with-idiom`. And responsibility of cleaning up after itself.

Let's discuss the problems in your code step by step...

**First up: [manage.py]**

```
def _get_api_instance():
    ...
    api_client = ApiClient(configuration=client_config)
    ...
    api_instance = IndexOperationsApi(api_client)
    return api_instance

...

# example call site, all call sites are completely broken...

def delete_index(name: str, timeout: int = None):
    ...
    api_instance = _get_api_instance()
    ...
```

Here, no cleanup possible of the connection pool, and no opportunity for connection reuse, hence it's slow as sludge
because of all the connection setup repeated again and again! Every call to any management API creates a new Client, and by implication a new connection pool, and by implication a single solitary connection. Which you never cleanup and results in a leaked file descriptor.

At the OS level, this means you have two leaked semaphores, AND two leaked pipes, AND one leaked socket. Five file descriptor leaks per API call. That's a lot of file descriptors leaked.

The proper choice would have been to change `manage.py` into a proper management API that takes a `ApiClient` as a constructor argument to the ManagementApi object, thus users can reuse the underlying pools. Not much different than `IndexOperationsApi` which does:

```
    def __init__(self, api_client=None):
        if api_client is None:
            api_client = ApiClient() # note this is a leak in all your existing APIs too!!!!
        self.api_client = api_client
```

**Second up: [IndexOperationsApi,  VectorOperationsApi]**

Both APIs suffer the same issue; line numbers listed at the end of this report for where the leaks occur. But in a nutshell, the APIs permissively take an `api_client` as a constructor argument (good) but falls back to creating one if it wasn't provided (bad).

Why bad? Because the constructor defines OWNERSHIP of the objects it manages, in this case `ApiClient`. And the *OperationsApi* takes no steps to do any cleanup, hence leaks, if it did create the `ApiClient` object.

My suggestion would be to take away the `is None` block of code that creates an `ApiClient` on behalf of users, and demand that all users create the client, thus reused. As an alternative, introduce a new class property, named `owns_client`, and add `__enter__`, `__exit__`, and `close` methods. In the close method, if `owns_client` is `True`, then call close on the client. You know what to do in `__enter__` and `__exit__`, it should be plainly obvious (implement the `with-idiom`).

This will fix the leaks in these two APIs.

**Detailed List of Leaks**

- `manage.py` (5 leaks at line 56)
- `IndexOperationsApi` (5 leaks at line 43)
- `VectorOperationsApi `(5 leaks at line 50)

### Expected Behavior

The product should NOT leak file descriptors.

### Steps To Reproduce

Create a directory of 64000 single word documents. Ingest them concurrently (in parallel if possible, to speed things up), watch the python client fail (die/crash) because of leaked file descriptors.

You can monitor file descriptors in real time:

```
#!/bin/bash

# Get the PID of the process
PID=$(pgrep -f "/Library/Frameworks/Python.framework/Versions/3.8/Resources/Python.app/Contents/MacOS/Python /Users/robert.buck/ws/workspaces/library-import/src/testing/venv/bin/driver crush -c 1000")

# Continuously monitor the open file count
while true; do
    echo "Open files for PID $PID :"
    lsof -p $PID
    open_file_count=$(lsof -p $PID | wc -l | awk '{print $1}')
    echo "Open file count for PID $PID: $open_file_count"
    sleep 5  # Adjust the interval as needed
done
```

### Relevant log output

```shell
Similar to:


.... 10-20 thousand more above...
Python 60130 robert.buck 7221u IPv4 0xd946e14d9f16b61b 0t0 TCP 192.168.0.103:60745->ec2-18-213-200-10.compute-1.amazonaws.com:https (ESTABLISHED)
Python 60130 robert.buck 7222 PIPE 0x2727afc4fb51e58a 16384 ->0xd97beed562c859f2
Python 60130 robert.buck 7223 PIPE 0xd97beed562c859f2 16384 ->0x2727afc4fb51e58a
Python 60130 robert.buck 7224r PSXSEM 0t0 /mp-r4wqhiyk
Python 60130 robert.buck 7225r PSXSEM 0t0 /mp-vpwtcej6
.... 10-20 thousand more below...
```

And...

```
WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa98a427af0>: Failed to establish a new connection: [Errno 16] Device or resource busy')': /vectors/upsert
WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(OSError(24, 'Too many open files'))': /vectors/upsert
WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa98a427310>: Failed to establish a new connection: [Errno 16] Device or resource busy')': /vectors/upsert
WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa98a427e20>: Failed to establish a new connection: [Errno 24] Too many open files')': /vectors/upsert
WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa98a4270a0>: Failed to establish a new connection: [Errno 24] Too many open files')': /vectors/upsert
WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa98a4276a0>: Failed to establish a new connection: [Errno 24] Too many open files')': /vectors/upsert
WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa97fc98880>: Failed to establish a new connection: [Errno 16] Device or resource busy')': /vectors/upsert
WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fa98dc96ac0>: Failed to establish a new connection: [Errno 16] Device or resource busy')': /vectors/upsert
WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'SSLError(OSError(24, 'Too many open files'))': /vectors/upsertINFO:root:Successfully shutdown the libtool import.Traceback (most recent call last):  File "/usr/local/lib/python3.8/site-packages/urllib3/util/ssl_.py", line 446, in ssl_wrap_socket    context.load_verify_locations(ca_certs, ca_cert_dir, ca_cert_data)OSError: [Errno 24] Too many open files
The above exception was the direct cause of the following exception:
Traceback (most recent call last):  

File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 791, in urlopen    response = self._make_request(  File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 492, in _make_request    raise new_e  
File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 468, in _make_request    self._validate_conn(conn)  
File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 1097, in _validate_conn    conn.connect()  
File "/usr/local/lib/python3.8/site-packages/urllib3/connection.py", line 642, in connect    sock_and_verified = _ssl_wrap_socket_and_match_hostname(  
File "/usr/local/lib/python3.8/site-packages/urllib3/connection.py", line 783, in _ssl_wrap_socket_and_match_hostname    ssl_sock = ssl_wrap_socket(  
File "/usr/local/lib/python3.8/site-packages/urllib3/util/ssl_.py", line 448, in ssl_wrap_socket    raise SSLError(e) from eurllib3.exceptions.SSLError: [Errno 24] Too many open files

The above exception was the direct cause of the following exception:
Traceback (most recent call last):  

File "/usr/local/bin/libtool", line 11, in <module>    load_entry_point('libtool==1.0.0', 'console_scripts', 'libtool')()  
File "/usr/local/lib/python3.8/site-packages/libtool/__main__.py", line 71, in main    args.func(args, verbose)  
File "/usr/local/lib/python3.8/site-packages/libtool/command/load.py", line 8, in load_action    loader.run()  
File "/usr/local/lib/python3.8/site-packages/libtool/core/loader.py", line 119, in run    loop.run_until_complete(asyncio.gather(*writer_tasks, ))  
File "/usr/lib64/python3.8/asyncio/base_events.py", line 616, in run_until_complete    return future.result()  
File "/usr/local/lib/python3.8/site-packages/libtool/core/loader.py", line 74, in writer    await plugin_writer(self.config, record, metadata)  
File "/usr/local/lib/python3.8/site-packages/libtool/core/backend/pinecone/backend.py", line 48, in write    Pinecone.from_texts(  
File "/usr/local/lib/python3.8/site-packages/langchain/vectorstores/pinecone.py", line 416, in from_texts    pinecone.add_texts(  
File "/usr/local/lib/python3.8/site-packages/langchain/vectorstores/pinecone.py", line 149, in add_texts    [res.get() for res in async_res]  
File "/usr/local/lib/python3.8/site-packages/langchain/vectorstores/pinecone.py", line 149, in <listcomp>    [res.get() for res in async_res]  
File "/usr/lib64/python3.8/multiprocessing/pool.py", line 771, in get    raise self._value  File "/usr/lib64/python3.8/multiprocessing/pool.py", line 125, in worker    result = (True, func(*args, **kwds))  
File "/usr/local/lib/python3.8/site-packages/pinecone/core/client/api_client.py", line 200, in __call_api    response_data = self.request(  
File "/usr/local/lib/python3.8/site-packages/pinecone/core/client/api_client.py", line 459, in request    return self.rest_client.POST(url,  
File "/usr/local/lib/python3.8/site-packages/pinecone/core/client/rest.py", line 271, in POST    return self.request("POST", url,  
File "/usr/local/lib/python3.8/site-packages/pinecone/core/client/rest.py", line 157, in request    r = self.pool_manager.request(  
File "/usr/local/lib/python3.8/site-packages/urllib3/_request_methods.py", line 118, in request    return self.request_encode_body(  
File "/usr/local/lib/python3.8/site-packages/urllib3/_request_methods.py", line 217, in request_encode_body    return self.urlopen(method, url, **extra_kw)  
File "/usr/local/lib/python3.8/site-packages/urllib3/poolmanager.py", line 443, in urlopen    response = conn.urlopen(method, u.request_uri, **kw)  
File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 875, in urlopen    return self.urlopen(  
File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 875, in urlopen    return self.urlopen(  
File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 875, in urlopen    return self.urlopen(  
File "/usr/local/lib/python3.8/site-packages/urllib3/connectionpool.py", line 845, in urlopen    retries = retries.increment(  
File "/usr/local/lib/python3.8/site-packages/urllib3/util/retry.py", line 515, in increment    raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='m01-8113f38.svc.us-east-1-aws.pinecone.io', port=443): Max retries exceeded with url: /vectors/upsert (Caused by SSLError(OSError(24, 'Too many open files')))
-- 
```
```


### Environment

```markdown
- OS: Mac, Red Hat Enterprise Linux, Any
- Python: 3.8.2
- pinecone: 2.2.4
```

### Additional Context

Work-around, avoid the APIs that leak.

**Comment by anawishnoff** [comment]
As this issue is from last year, I'll go ahead and close this as stale. However please feel free to reopen or file a new issue if you're still having issues - thank you!
